modules/gloo/pkg/vsfactory/target.go

```diff
@@ -176,3 +180,57 @@ func formatHost(port, defaultPort int, hostname string) string {
 	}
 	return fmt.Sprintf("%s:%d", hostname, port)
 }
+
+// check if a candidate service is scale zero, does not check if the service CAN be scaled from zero
+// returns true if there are NO k8s endpoints for the service allocated yet
+func (f *VirtualServiceFactory) isScaleZero(appName, version, namespace string) (bool, error) {
+	// for services scaled to zero, ensure endpoint has an IP address
+	req, err := labels.NewRequirement("app", selection.Equals, []string{appName})
+	if err != nil {
+		return false, err
+	}
+	reqVer, err := labels.NewRequirement("version", selection.Equals, []string{version})
+	if err != nil {
+		return false, err
+	}
+	selector := labels.NewSelector().Add(*req)
+	selector = selector.Add(*reqVer)
+	// use a selector for the app + version endpoint (ie RioApp-v0) versus app endpoint to see if service is up
+	k8sEndpoints, err := f.endpoints.List(namespace, selector)
+	if err != nil {
+		return false, err
+	}
+	if len(k8sEndpoints) == 0 {
+		logrus.Debugf("no corev1.Endpoints found for %s with version %s", appName, version)
+		return false, nil
+	}
+	// check all k8s endpoints to see if any IP addresses exist
+	for _, endpoint := range k8sEndpoints {
+		if len(endpoint.Subsets) != 0 {
+			// ip address allocated for this service, should not be scale zero
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// determineHosts returns a set of hosts based upon the rio services endpoint URLs
+func determineHosts(hostNameEndpoints []string) ([]string, error) {
```

Comment:
@cbron Line [65-79](https://github.com/rancher/rio/pull/985/files#diff-0586357b98c09d1ed2a1132e44706c4dL65-L79) was moved into this function